src/m365/spo/commands/site/site-admin-add.ts

```diff
@@ -0,0 +1,358 @@
+          return `${args.options.userId} is not a valid GUID`;
+        }
+
+        if (args.options.userName && !validation.isValidUserPrincipalName(args.options.userName)) {
+          return `${args.options.userName} is not a valid userName`;
+        }
+
+        if (args.options.groupId &&
+          !validation.isValidGuid(args.options.groupId)) {
+          return `${args.options.groupId} is not a valid GUID`;
+        }
+
+        return validation.isValidSharePointUrl(args.options.siteUrl);
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push({ options: ['userId', 'userName', 'groupId', 'groupName'] });
+  }
+
+  #initTypes(): void {
+    this.types.string.push('siteUrl', 'userId', 'userName', 'groupId', 'groupName');
+    this.types.boolean.push('primary', 'asAdmin');
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    try {
+      const loginNameToAdd = await this.getCorrectLoginName(args.options);
+      if (args.options.asAdmin) {
+        await this.callActionAsAdmin(logger, args, loginNameToAdd);
+        return;
+      }
+
+      await this.callAction(logger, args, loginNameToAdd);
+    }
+    catch (err: any) {
+      this.handleRejectedODataJsonPromise(err);
+    }
+  }
+
+  private async callActionAsAdmin(logger: Logger, args: CommandArgs, loginNameToAdd: string): Promise<void> {
+    if (this.verbose) {
+      await logger.logToStderr('Adding site administrator as an administrator...');
+    }
+
+    const adminUrl: string = await spo.getSpoAdminUrl(logger, this.debug);
+    const siteId = await this.getSiteIdBasedOnUrl(args.options.siteUrl, logger);
+    const siteAdmins = (await this.getSiteAdmins(adminUrl, siteId)).map(u => u.loginName);
+    siteAdmins.push(loginNameToAdd);
+    await this.setSiteAdminsAsAdmin(adminUrl, siteId, siteAdmins);
+    if (args.options.primary) {
+      await this.setPrimaryAdminAsAdmin(adminUrl, siteId, loginNameToAdd);
+    }
+  }
+
+  private async getSiteIdBasedOnUrl(siteUrl: string, logger: Logger): Promise<string> {
+    const siteGraphId = await spo.getSiteId(siteUrl, logger, this.verbose);
+    const match = siteGraphId.match(/,([a-f0-9\-]{36}),/i);
+    if (!match) {
+      throw `Site with URL ${siteUrl} not found`;
+    }
+
+    return match[1];
+  }
+
+  private async getSiteAdmins(adminUrl: string, siteId: string): Promise<AdminUserResult[]> {
+    const requestOptions: CliRequestOptions = {
+      url: `${adminUrl}/_api/SPO.Tenant/GetSiteAdministrators?siteId='${siteId}'`,
+      headers: {
+        accept: 'application/json;odata=nometadata',
+        'content-type': 'application/json;charset=utf-8'
+      }
+    };
+
+    const response: string = await request.post<string>(requestOptions);
+    const responseContent: AdminResult = JSON.parse(response);
+    return responseContent.value;
+  }
+
+  private async getCorrectLoginName(options: Options): Promise<string> {
+    if (options.userId || options.userName) {
```

Comment:
Let's make use of the utils for this. (entraUser)

Reply:
ok here I used entraUser.getUpnByUserId, I change the behavior when we use userName